src/Shared/ArgumentEscaper.cs

```diff
@@ -0,0 +1,128 @@
+        ///
+        /// See here for more info:
+        /// http://blogs.msdn.com/b/twistylittlepassagesallalike/archive/2011/04/23/everyone-quotes-arguments-the-wrong-way.aspx
+        /// </summary>
+        /// <param name="args"></param>
+        /// <returns></returns>
+        private static IEnumerable<string> EscapeArgArray(IEnumerable<string> args)
+        {
+            var escapedArgs = new List<string>();
+
+            foreach (var arg in args)
+            {
+                escapedArgs.Add(EscapeSingleArg(arg));
+            }
+
+            return escapedArgs;
+        }
+
+        public static string EscapeSingleArg(string arg)
+        {
+            var sb = new StringBuilder();
+
+            var needsQuotes = ShouldSurroundWithQuotes(arg);
+            var isQuoted = needsQuotes || IsSurroundedWithQuotes(arg);
+
+            if (needsQuotes) sb.Append("\"");
+
+            for (int i = 0; i < arg.Length; ++i)
+            {
+                var backslashCount = 0;
+
+                // Consume All Backslashes
+                while (i < arg.Length && arg[i] == '\\')
+                {
+                    backslashCount++;
+                    i++;
+                }
+
+                // Escape any backslashes at the end of the arg
+                // when the argument is also quoted.
+                // This ensures the outside quote is interpreted as
+                // an argument delimiter
+                if (i == arg.Length && isQuoted)
+                {
+                    sb.Append('\\', 2 * backslashCount);
+                }
+
+                // At then end of the arg, which isn't quoted,
+                // just add the backslashes, no need to escape
+                else if (i == arg.Length)
+                {
+                    sb.Append('\\', backslashCount);
+                }
+
+                // Escape any preceding backslashes and the quote
+                else if (arg[i] == '"')
+                {
+                    sb.Append('\\', (2 * backslashCount) + 1);
+                    sb.Append('"');
+                }
+
+                // Output any consumed backslashes and the character
+                else
+                {
+                    sb.Append('\\', backslashCount);
+                    sb.Append(arg[i]);
+                }
+            }
+
+            if (needsQuotes) sb.Append("\"");
+
+            return sb.ToString();
+        }
+
+        internal static bool ShouldSurroundWithQuotes(string argument)
+        {
+            // Don't quote already quoted strings
+            if (IsSurroundedWithQuotes(argument))
+            {
+                return false;
+            }
+
+            // Only quote if whitespace exists in the string
+            return ArgumentContainsWhitespace(argument);
+        }
+
+        internal static bool IsSurroundedWithQuotes(string argument)
+        {
+            return argument.StartsWith("\"", StringComparison.Ordinal) &&
```

Comment:
Is there anything particular to `\` that you need `StringComparison.Orginal` here? Since this is single char check you could just index the string and check what's there.